### PR TITLE
Added unit test for tcp server

### PR DESCRIPTION
**Pull request by urmahp**
Added a check that the thread is running before joining.

Made it possible to reuse the port, after the server has been closed.

Added a check that the callback functions has been configured before starting the worker thread.

### CODE REVIEW

File: src/comm/tcp_server.cpp [hunk 1]
```diff
@@ -288,11 +294,20 @@ void TCPServer::worker()
   URCL_LOG_DEBUG("Finished worker thread of TCPServer");
 }
 
-void TCPServer::start()
+bool TCPServer::start()
 {
-  URCL_LOG_DEBUG("Starting worker thread");
-  keep_running_ = true;
-  worker_thread_ = std::thread(&TCPServer::worker, this);
+  if (new_connection_callback_ && disconnect_callback_ && message_callback_)
```
Comment (fmauch) [inline src/comm/tcp_server.cpp]: Actually, that was on purpose that those could be empty. This way, the user doesn't **have** to provide the callbacks.

However, that should probably be checked when they are called. I didn't write the checks as it did work without them. I am still curious why it does, though...

I'll update my branch accordingly.

File: src/comm/tcp_server.cpp [hunk 2]
```diff
@@ -62,6 +62,9 @@ void TCPServer::init()
   {
     throw std::system_error(std::error_code(errno, std::generic_category()), "Failed to create socket endpoint");
   }
+  int flag = 1;
+  setsockopt(listen_fd_, SOL_SOCKET, SO_REUSEADDR, &flag, sizeof(int));
```
Comment (fmauch) [inline src/comm/tcp_server.cpp:66]: fine with me :-)

File: tests/test_tcp_server.cpp [hunk 3]
```diff
@@ -0,0 +1,341 @@
...
+    message_ = std::string(buffer);
+    message_cv_.notify_one();
+    message_callback_ = true;
+  }
+
+  bool waitForConnectionCallback(int milliseconds = 100)
+  {
+    std::unique_lock<std::mutex> lk(connect_mutex_);
+    auto now = std::chrono::system_clock::now();
+    if (connect_cv_.wait_until(lk, now + std::chrono::milliseconds(milliseconds)) == std::cv_status::no_timeout ||
```
Comment (fmauch) [inline tests/test_tcp_server.cpp]:
```suggestion
    if (connect_cv_.wait_for(lk, std::chrono::milliseconds(milliseconds)) == std::cv_status::no_timeout ||
```

File: tests/test_tcp_server.cpp [hunk 4]
```diff
@@ -0,0 +1,341 @@
...
+    {
+      return false;
+    }
+  }
+
+  bool waitForDisconnectionCallback(int milliseconds = 100)
+  {
+    std::unique_lock<std::mutex> lk(disconnect_mutex_);
+    auto now = std::chrono::system_clock::now();
+    if (disconnect_cv_.wait_until(lk, now + std::chrono::milliseconds(milliseconds)) == std::cv_status::no_timeout ||
```
Comment (fmauch) [inline tests/test_tcp_server.cpp]:
```suggestion
    if (disconnect_cv_.wait_for(lk, std::chrono::milliseconds(milliseconds)) == std::cv_status::no_timeout ||
```

File: tests/test_tcp_server.cpp [hunk 5]
```diff
@@ -0,0 +1,341 @@
...
+    {
+      return false;
+    }
+  }
+
+  bool waitForMessageCallback(int milliseconds = 100)
+  {
+    std::unique_lock<std::mutex> lk(message_mutex_);
+    auto now = std::chrono::system_clock::now();
+    if (message_cv_.wait_until(lk, now + std::chrono::milliseconds(milliseconds)) == std::cv_status::no_timeout ||
```
Comment (fmauch) [inline tests/test_tcp_server.cpp]:
```suggestion
    if (message_cv_.wait_for(lk, std::chrono::milliseconds(milliseconds)) == std::cv_status::no_timeout ||
```

File: tests/test_tcp_server.cpp [hunk 6]
```diff
@@ -0,0 +1,341 @@
...
+TEST_F(TCPServerTest, socket_creation)
+{
+  comm::TCPServer server(port_);
+
+  // Shouldn't be able to create antoher server on same port
+  EXPECT_THROW(comm::TCPServer server2(port_), std::system_error);
+
+  // Shouldn't be able to start the worker thread, when no callbacks has been registered
+  EXPECT_FALSE(server.start());
+
```
Comment (fmauch) [inline tests/test_tcp_server.cpp]: I suggest removing this check as reasoned above.

File: tests/test_tcp_server.cpp [hunk 7]
```diff
@@ -0,0 +1,341 @@
...
+  comm::TCPServer server(port_);
+  server.setMessageCallback(std::bind(&TCPServerTest_unlimited_clients_allowed_Test::messageCallback, this,
+                                      std::placeholders::_1, std::placeholders::_2));
+  server.setConnectCallback(
+      std::bind(&TCPServerTest_unlimited_clients_allowed_Test::connectionCallback, this, std::placeholders::_1));
+  server.setDisconnectCallback(
+      std::bind(&TCPServerTest_unlimited_clients_allowed_Test::disconnectionCallback, this, std::placeholders::_1));
+  server.start();
+
+  // Test that unlimited number of clients can connect
```
Comment (fmauch) [inline tests/test_tcp_server.cpp]: 3 == unlimited? :laughing:

**Review by fmauch — COMMENTED** [review]
This is looking very good!

I will update my branch regarding the empty functions then we can adapt this accordingly and merge it!

File: src/comm/tcp_server.cpp (hunk 1 above)
Comment (fmauch) [reply src/comm/tcp_server.cpp]: See faffde1

File: tests/test_tcp_server.cpp (hunk 7 above)
Comment (urmahp) [reply tests/test_tcp_server.cpp]: I will change the comment and the test name :laughing:

**Review by fmauch — APPROVED** [review]
Looks good to me. Thank you @urmahp!

**Comment by fmauch** [comment]
Merging this despite the failing link checks. I checked the links manually and they are valid. There seem to be issues with the w3.org servers.

**Pull request merged**